### PR TITLE
Improve spacefinder visualiser

### PR DESCRIPTION
**Pull request by simonbyford**
## What does this change?

Makes a few improvements to the spacefinder visualiser.

Firstly, if a candidate paragraph is too close to multiple DOM elements, it now highlights them all. Previously, it only highlighted one, which could be misleading.

Secondly, when highlighting elements, the tool now overlays information about the "required" and "actual" distances between the candidate paragraph and the given element. For example, if a candidate paragraph is 40px from a certain element, but it must be at least 100px, the overlay will read "40px/100px".

Thirdly, it highlights in red because it looks better.

## Performance

The first improvement (highlighting all elements) required a tweak to the way spacefinder works. Previously, when looping through elements to check if they are too close, it stopped when it found one. This makes sense and was likely for performance reasons. Now it exhaustively checks every possible element, even after it finds one.

This has a performance impact. I have attempted to benchmark this by running spacefinder on 3 articles of different complexity before and after the change. I found that it approximately doubled the time taken to run `enforceRules` for `inline1` ads. For a very complex article like [this one](https://www.theguardian.com/news/2022/feb/20/credit-suisse-secrets-leak-unmasks-criminals-fraudsters-corrupt-politicians) this amounts to an increase of approximately 70ms to 150ms. The full results are [here](https://docs.google.com/spreadsheets/d/1lFXB-fPWzja8XPZgCZIo0KKsraFYXe8fijFbLSQw9ZA/edit?usp=sharing).

Our options are therefore:

1) accept this small performance hit
2) rewrite the function to behave differently depending on whether we're in debug mode or not (i.e. by accepting an `exhaustive` parameter, for example)
3) abandon this change

I think it's a worthwhile change so would lean towards 1 or 2.
 
## Screenshots

Apologies for the painfully slow videos, something happened when I converted them to mp4.

### Before

https://user-images.githubusercontent.com/7423751/157011826-d705bdf9-576e-4b31-844a-3a38606b2509.mp4

### After

https://user-images.githubusercontent.com/7423751/157013526-598f13d4-d033-48ef-989e-b69e73a2cba4.mp4

## What is the value of this and can you measure success?

The tool is more useful and enables us to find bugs more easily.

### CODE REVIEW

**Review by mxdvl — COMMENTED** [review]
Are we shipping the debug code to all users, or is there a way to ensure we omit it from the bundle if the debug parameters is absent? As this piece of code grows, it might become more relevant to do so.

> Apologies for the painfully slow videos, something happened when I converted them to mp4.

You can pop your `.mov` in Github directly!

File: static/src/javascripts/projects/common/modules/spacefinder.js [hunk 1]
```diff
@@ -173,6 +186,7 @@ const enforceRules = (measurements, rules, exclusions) => {
 
 	// enforce selector rules
 	if (rules.selectors) {
+		let selectorExclusions = [];
```
Comment (chrislomaxjones) [inline static/src/javascripts/projects/common/modules/spacefinder.js:189]: Can this be a `const` here? I understand that the contents of the array are being mutated, but the reference of the array itself is never reassigned?

File: static/src/javascripts/projects/common/modules/spacefinder.js [hunk 2]
```diff
@@ -124,15 +124,28 @@ const testCandidate = (rule, candidate, opponent) => {
 	const pass = isMinAbove || isMinBelow;
 
 	if (!pass) {
-		candidate.meta.tooClose.push(opponent.element);
+		// if the test fails, add debug information to the candidate metadata
+		let isBelow = candidate.top < opponent.top;
+		let required = isBelow ? rule.minBelow : rule.minAbove;
+		let actual = isBelow
```
Comment (chrislomaxjones) [inline static/src/javascripts/projects/common/modules/spacefinder.js:130]: Can these be be `const` as well?

File: static/src/javascripts/projects/common/modules/spacefinder.js (hunk 1 above)
Comment (simonbyford) [reply static/src/javascripts/projects/common/modules/spacefinder.js]: Yup you're right! Have fixed

**Review by chrislomaxjones — APPROVED** [review]
Nice improvement!

**Comment by simonbyford** [comment]
> Are we shipping the debug code to all users, or is there a way to ensure we omit it from the bundle if the debug parameters is absent?

Right now we're shipping this to all users. but this is something we could consider in future

> You can pop your .mov in Github directly!

The chrome extension I use for screen recording generates `.webm` files annoyingly!

**Comment by lucymonie** [comment]
Coming to this late! I agree with your suggestion
> rewrite the function to behave differently depending on whether we're in debug mode or not (i.e. by accepting an exhaustive parameter, for example)

Maybe it would be good to have a trello card for this change?

**Pull request merged**